### PR TITLE
add v8 checksum calculation logic (DAT-14888)

### PR DESCRIPTION
**Pull request by StevenMassaro**
## Impact

- [ ] Bug fix (non-breaking change which fixes expected existing functionality)
- [ ] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Things to be aware of

The initial work to switch from v8 to v9 checksum calc logic is here: https://github.com/liquibase/liquibase/pull/3914. I recommend that you have this open while reviewing this code to confirm that the original v8 logic was ported correctly here.

## Things to worry about

- [ ] Do we need to increment to v10 with this release? This needs to be discussed.

### CODE REVIEW

File: liquibase-standard/src/main/java/liquibase/change/DatabaseChangeProperties.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package liquibase.change;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.METHOD})
+public @interface DatabaseChangeProperties {
```
Comment (StevenMassaro) [inline liquibase-standard/src/main/java/liquibase/change/DatabaseChangeProperties.java:10]: Javadoc

File: liquibase-standard/src/main/java/liquibase/change/core/CreateProcedureChange.java [hunk 2]
```diff
@@ -297,7 +356,8 @@ public String[] getExcludedFieldFilters() {
                 "triggerBody",
                 "functionBody",
                 "packageText",
-                "packageBodyText"
+                "packageBodyText"//,
+                //"procedureBody"
```
Comment (StevenMassaro) [inline liquibase-standard/src/main/java/liquibase/change/core/CreateProcedureChange.java:360]: ?

File: liquibase-standard/src/main/java/liquibase/change/core/CreateProcedureChange.java [hunk 3]
```diff
@@ -286,7 +340,12 @@ protected CheckSum generateCheckSum(String sqlText, ChecksumVersions version) {
      * change types.
      */
     @Override
-    public String[] getExcludedFieldFilters() {
+    public String[] getExcludedFieldFilters(ChecksumVersions version) {
+        if (version == ChecksumVersions.V8) {
+            return new String[]{
+                    "procedureText"
+            };
+        }
```
Comment (StevenMassaro) [inline liquibase-standard/src/main/java/liquibase/change/core/CreateProcedureChange.java:348]: Is this necessary?

File: liquibase-standard/src/test/groovy/liquibase/blah.groovy [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+import liquibase.resource.ClassLoaderResourceAccessor;
+import liquibase.statement.SqlStatement
+import liquibase.statement.core.InsertOrUpdateStatement;
+import liquibase.test.JUnitResourceAccessor
+import liquibase.database.core.MSSQLDatabase
+import spock.lang.Unroll
+
+import static org.junit.Assert.*
+
+public class LoadUpdateDataChangeTest2 extends StandardChangeTest {
```
Comment (StevenMassaro) [inline liquibase-standard/src/test/groovy/liquibase/blah.groovy:19]: Delete this

File: liquibase-standard/src/main/java/liquibase/change/AbstractChange.java [hunk 5]
```diff
@@ -490,21 +517,28 @@ public boolean supportsRollback(Database database) {
      */
     @Override
     public CheckSum generateCheckSum(ChecksumVersions version) {
-        if (version.equals(ChecksumVersions.V8)) {
-            return CheckSum.compute(new StringChangeLogSerializer().serialize(this, false), version);
+        try {
+            return Scope.child(Collections.singletonMap(Scope.Attr.currentChecksumVersion.toString(), version), () -> internalGenerateChecksum(version));
+        } catch (Exception e) {
+            // todo logging?
+            // todo is this how we should handle it?
```
Comment (StevenMassaro) [inline liquibase-standard/src/main/java/liquibase/change/AbstractChange.java:524]: Resolve

File: liquibase-standard/src/main/java/liquibase/change/AbstractChange.java [hunk 6]
```diff
@@ -107,6 +109,31 @@ public ChangeMetaData createChangeMetaData() {
         }
     }
 
+    private DatabaseChangeProperty findDatabaseChangePropertyAnnotationMatchingCurrentChecksumVersion(Method readMethod) {
```
Comment (StevenMassaro) [inline liquibase-standard/src/main/java/liquibase/change/AbstractChange.java:112]: Javadoc

File: liquibase-standard/src/main/java/liquibase/change/core/CreateProcedureChange.java [hunk 7]
```diff
@@ -240,10 +246,58 @@ public InputStream openSqlStream() throws IOException {
...
-        return generateCheckSum(this.procedureText, version);
+        if (version == ChecksumVersions.V8) {
+            return generateCheckSumV8(version);
+        }
+        return generateCheckSumV9(this.procedureText, version);
     }
 
-    protected CheckSum generateCheckSum(String sqlText, ChecksumVersions version) {
+    @Deprecated
+    protected CheckSum generateCheckSumV8(ChecksumVersions version) {
```
Comment (StevenMassaro) [inline liquibase-standard/src/main/java/liquibase/change/core/CreateProcedureChange.java:256]: Make private

File: liquibase-standard/src/main/java/liquibase/change/AbstractChange.java [hunk 8]
```diff
@@ -107,6 +109,31 @@ public ChangeMetaData createChangeMetaData() {
...
+        DatabaseChangeProperty[] annotations = readMethod.getAnnotationsByType(DatabaseChangeProperty.class);
+        if (annotations.length == 1) {
+            return annotations[0];
+        } else if (annotations.length > 1) {
+            // todo - need some kind of check - maybe a test- that no one duplicates versions of the annotation anywhere
+            // todo put the checksum version in the scope
```
Comment (StevenMassaro) [inline liquibase-standard/src/main/java/liquibase/change/AbstractChange.java:118]: Resolve

File: liquibase-standard/src/main/java/liquibase/change/AbstractSQLChange.java [hunk 9]
```diff
@@ -351,4 +357,117 @@ public void close() throws IOException {
             stream.close();
         }
     }
+
```
Comment (filipelautert) [inline liquibase-standard/src/main/java/liquibase/change/AbstractSQLChange.java:360]: Hm... would not be better to have this NormalizingStreamV8 class in a separated file? It's so big .

File: liquibase-standard/src/main/java/liquibase/change/core/CreateProcedureChange.java [hunk 10]
```diff
@@ -240,10 +246,58 @@ public InputStream openSqlStream() throws IOException {
      */
     @Override
     public CheckSum generateCheckSum(ChecksumVersions version) {
-        return generateCheckSum(this.procedureText, version);
+        if (version == ChecksumVersions.V8) {
+            return generateCheckSumV8(version);
+        }
+        return generateCheckSumV9(this.procedureText, version);
```
Comment (filipelautert) [inline liquibase-standard/src/main/java/liquibase/change/core/CreateProcedureChange.java:252]: Maybe generateCheckSumV9 should be called just "generateCheckSum()" or "generateCheckSumLatest()"? As when we release checksum v10 we may not change all those methods, then it will be generating v9 and v10 .

File: liquibase-standard/src/main/java/liquibase/change/core/CreateViewChange.java [hunk 11]
```diff
@@ -203,6 +207,13 @@ public String[] getExcludedFieldFilters() {
      */
     @Override
     public CheckSum generateCheckSum(ChecksumVersions version) {
+        if (version == ChecksumVersions.V8) {
+            return generateCheckSumV8(version);
+        }
+        return generateCheckSumV9(version);
```
Comment (filipelautert) [inline liquibase-standard/src/main/java/liquibase/change/core/CreateViewChange.java:213]: Same as above.

**Review by filipelautert — COMMENTED** [review]
Mostly cosmetic changes

**Review by filipelautert — APPROVED** [review]
